examples/miscellaneous/plot_outlier_detection_bench.py

```diff
@@ -17,181 +19,361 @@
-        X = X[idx]  # reduce the sample size
-        y = y[idx]
-
-        # inliers are those with attribute 2
-        # outliers are those with attribute 4
-        s = (y == 2) + (y == 4)
-        X = X[s, :]
-        y = y[s]
-        y = (y != 2).astype(int)
-    if dataset_name in ["glass", "wdbc", "cardiotocography"]:
-        dataset = fetch_openml(
-            name=dataset_name, version=1, as_frame=False, parser="pandas"
-        )
-        X = dataset.data
-        y = dataset.target
-
-        if dataset_name == "glass":
-            s = y == "tableware"
-            y = s.astype(int)
-        if dataset_name == "wdbc":
-            s = y == "2"
-            y = s.astype(int)
-            X_mal, y_mal = X[s], y[s]
-            X_ben, y_ben = X[~s], y[~s]
-
-            # downsampled to 39 points (9.8% outliers)
-            idx = rng.choice(y_mal.shape[0], 39, replace=False)
-            X_mal2 = X_mal[idx]
-            y_mal2 = y_mal[idx]
-            X = np.concatenate((X_ben, X_mal2), axis=0)
-            y = np.concatenate((y_ben, y_mal2), axis=0)
-        if dataset_name == "cardiotocography":
-            s = y == "3"
-            y = s.astype(int)
-    # 0 represents inliers, and 1 represents outliers
-    y = pd.Series(y, dtype="category")
-    return (X, y)
-
-
-# %%
-# Define an outlier prediction function
-# -------------------------------------
-# There is no particular reason to choose algorithms
-# :class:`~sklearn.neighbors.LocalOutlierFactor` and
-# :class:`~sklearn.ensemble.IsolationForest`. The goal is to show that
-# different algorithm performs well on different datasets. The following
-# `compute_prediction` function returns average outlier score of X.
+n_samples, anomaly_frac = X.shape[0], y.mean()
+print(f"{n_samples} datapoints with anomaly propotion of {anomaly_frac:.02%}")
 
+# %%
+# The SA dataset contains 41 features out of which 3 are categorical:
+# "protocol_type", "service" and "flag".
 
-from sklearn.neighbors import LocalOutlierFactor
-from sklearn.ensemble import IsolationForest
+# %%
+y_true = {}
+y_pred = {"LOF": {}, "IForest": {}}
+model_names = ["LOF", "IForest"]
+cat_columns = ["protocol_type", "service", "flag"]
+
+y_true["KDDCup99 - SA"] = y
+for model_name in model_names:
+    y_pred[model_name]["KDDCup99 - SA"] = fit_predict(
+        X,
+        model_name=model_name,
+        categorical_columns=cat_columns,
+        n_neighbors=int(n_samples * anomaly_frac),
+    )
 
+# %%
+# In this example we set `n_neighbors` to match the number of anomalies on the
+# dataset. This is a good heuristic as long as the proportion of outliers is not
+# very low. Notice that this means the optimal number of neighbors scales with
+# the number of samples. Therefore the fit time of LOF models increases linearly
+# with the size of the dataset. If one had access to ground truth labels and was
+# to additionally tune the number of neighbors, the whole computation would
+# result quadratic on `n_samples`.
```

Comment:
Did some words go missing here? Somehow it reads like the sentence is missing some words.

Reply:
In a real life discussion with @jjerphan we realized that the statement about the fit time is not exactly accurate. I decided to get rid of that part and instead better explain what I meant for "as long as the proportion of outliers is not very low".